### ISSUE TITLE
Shortcuts works only in current keyboard layout

### ISSUE THREAD

**Issue by bitbybit** [IgnorantGuru/spacefm#57]
Ctrl + C / Ctrl + V do not work in russian keyboard layout
Ctrl + F — search instead of "New folder"

When trying to set new shortcut with russian layout then it works only in russian layout

Arch Linux. Openbox.

**Comment by IgnorantGuru** [comment]
You might also have a look at the [emit-key](http://ignorantguru.github.io/spacefm/spacefm-manual-en.html#sockets-methods-emit-key) socket command, which allows you to create a menu item which activates another.  This would allow you to have two (or more) key shortcuts which activate the same item.

**Comment by Vladimir-csp** [comment]
That would be impractical.
How do most of the other software allow shortcuts to work regardless of layout? It seems they bind to a key, not a symbol it produces.

**Comment by IgnorantGuru** [comment]
Well I don't mind telling you what little I know of it.  The code I've seen is pretty hackish, transforming some keycodes into others, basically assigning multiple keycodes to a single action, with lots of exceptions, etc.  If custom key shortcuts in other software work on multiple layouts then I think they must pass the keycodes through a filter.

The GTK event handler gives SpaceFM the keycodes, which are associated with actions.  If you want multiple layouts to work, I would say your best bet is to provide me with code - for example, a function that transforms your layout's keycodes into Latin keycodes.  I haven't come across anything like this, like I said most of what I've seen looked hacked.  But I haven't gone looking for it either - it probably exists.  I have no objection to this improvement so long as it doesn't reduce the flexibility of using any key, but it's not high on the list of priorities, and SpaceFM is in a slow development phase besides.  If you want this done, providing research or ready-to-use functions would be the way to encourage it.  If you think SpaceFM needs to pass keycodes through a filter, what's the filter?  Find the answers.

Fyi, SpaceFM bypasses GTK's normal methods for maintaining key shortcuts.  It only uses the accelerators to display the key shortcut in the menu.  SpaceFM's method is a prototype system that was hacked in to extend legacy PCManFM's flexibility (considerably) with minimal effort.  That may be why it lacks the refinement of other methods.

Believe it or not, I've spent thousands of hours developing SpaceFM.  These things don't just happen by themselves, and a project of this size is not usually developed by (mostly) one person.  If people want more functionality, they can whine and complain or they can get things done.  Take your pick.  This is a large application with MANY potential directions of development, and no lack of good ideas.  Even when people do want to contribute things, checking, fixing, integrating, documenting, releasing, and supporting their work is more work for me.  So if you want changes included upstream, make your work or research as complete as possible, requiring little action.  Otherwise you're just adding to a very long TODO list.  No matter how great and wonderful and important your idea is, there are LOTS of those on the list, even more than you see listed in the hundreds of requests here on github.

You claim I "reject" this issue but that's not true at all - it's open.  It doesn't have a high priority.  If you think you will motivate me to work on it by complaining or shaming you are mistaken.  I get that all the time from people trying to tell me how to spend my free time, and I know just how to handle them.

All of that said, I appreciate your positive contributions to the project.  With your level of interest, you may want to get more into detailed research or coding contributions for some of these well-contained issues.  I come up with good ideas all the time for SpaceFM, and like many of your ideas, they end up on a long list.

**Comment by Vladimir-csp** [comment]
Sorry about 'reject', I made a quick reply without rechecking while some things in my memory got messed up.

**Comment by IgnorantGuru** [comment]
Taking a quick look at the old PCManFM-Mod code in ptk-file-browser.c, this is how it looks like it was done.  This does look like it might be portable to SpaceFM without much work.

~~~~

    // Make key bindings work when the current keyboard layout is not latin
    if ( modifier != 0 && !is_latin_shortcut_key( event->keyval ) ) {
        // We have a non-latin char, try other keyboard groups
        GdkKeymapKey *keys;
        guint *keyvals;
        gint n_entries;
        gint level;

        if ( gdk_keymap_translate_keyboard_state( NULL,
                                                  event->hardware_keycode,
                                                  (GdkModifierType)event->state,
                                                  event->group,
                                                  NULL, NULL, &level, NULL )
            && gdk_keymap_get_entries_for_keycode( NULL,
                                                   event->hardware_keycode,
                                                   &keys, &keyvals,
                                                   &n_entries ) ) {
            gint n;
            for ( n=0; n<n_entries; n++ ) {
                if ( keys[n].group == event->group ) {
                    // Skip keys from the same group
                    continue;
                }
                if ( keys[n].level != level ) {
                    // Allow only same level keys
                    continue;
                }
                if ( is_latin_shortcut_key( keyvals[n] ) ) {
                    // Latin character found
                    event->keyval = keyvals[n];
                    break;
                }
            }
            g_free( keys );
            g_free( keyvals );
        }
    }
~~~~

**Comment by IgnorantGuru** [comment]
What will need to be done with this (mostly notes to myself):

* write a non_latin_key_transpose function based on above
* insert this function into all key event handlers in SpaceFM
* write a function for configuration upgrade to transpose  existing custom non-latin key shortcuts.  This will be a one-way process so downgrading will render those shortcuts inoperable
* update documentation

**Comment by Vladimir-csp** [comment]
Thank you!
Maybe you`ve already seen it, but here is a summary of different toolkits` algorithms dealing with this issue: http://helgo.net/simon/hacking/shortcut.html

**Comment by IgnorantGuru** [comment]
The code above will probably make it 'mostly work', but I doubt it will be perfect, as that discussion shows.  If anyone wants to suggest another function before I get to this, you might want to explore the options, keeping in mind no additional deps.  I'll refactor it in a way that the transpose function can be easily replaced anyway, not spread all over.  However, changing the function later may be undesirable because it may break custom key shortcuts.

You can see how even seemingly small things balloon into sub-projects once you get into details.  Nothing is ever easy, but getting at least some basic support for this doesn't look too bad.  I had remembered seeing more hackish stuff elsewhere in PCManFM (such as how it recognized Ctrl+C), but that function above looks general enough for this use.

**Comment by IgnorantGuru** [comment]
I have created a temporary 'key' branch that roughly [transposes non-latin keypresses](https://github.com/IgnorantGuru/spacefm/commit/6b698838d4ed46876c04568045e2e4a4c055e5c6) into their latin equivalent.  This occurs before the keypresses are compared to key shortcuts.

In the Set Key dialog, if you press a non-latin key, it should transpose it and also show the original keycode in square brackets.

Please test the key branch using the BUILD NEXT instructions in README, substituting 'key' for 'next' in those instructions.

Note that this branch does not yet convert key shortcuts stored in your config.  I'm mostly interested in whether default shortcuts like Ctrl+C now work for other layouts, and whether new shortcuts you add work consistently across all the layouts you use.

If this tests okay, I will then add a conversion when upgrading to 1.0.3 which converts custom non-latin key shortcuts into their latin equivalent using the same function.  This will make downgrading break key shortcuts.  (The only other option would be to continue storing non-latin keycodes and converting them upon comparison, but that would mean converting the keypress as well as all the key shortcuts to be compared, which is too inefficient.)

So this is just an initial commit to see how the transpose function is performing, and whether this enables default key shortcuts like Ctrl+C.

**If you use a non-latin keyboard layout, this is your chance to test this function**.  Once it's added in the next version, changing the function will mean further breakage for downgrades, so I will resist changing it much.  So testing now is strongly recommended.

**Comment by Vladimir-csp** [comment]
Great!
Basic shortcuts seem to work. There is an issue with search-as-you-type: the first letter typed is latin.
I've installed key branch for everyday use, so more testing will follow.

**Comment by IgnorantGuru** [comment]
Thanks, please see if eb9b075 corrects the find-as-you-type search.

**Comment by IgnorantGuru** [comment]
Actually you can wait on testing that last commit as I don't think that will fix search for icon view at least.  I'll update it soon.

**Comment by IgnorantGuru** [comment]
This is ready for testing.

**Comment by Vladimir-csp** [comment]
So far so good...

**Comment by Vladimir-csp** [comment]
It fails to translate these: 
```
ru: х ъ ж э б ю . ё
en: [ ] ; ' , . / `
```
But if it is too impractical to fix, I can live without it.

**Comment by IgnorantGuru** [comment]
> It fails to translate these: 

Such conversions probably differ between layouts, so if a fix was added for that, it would probably be specific to an event->group in [transpose_nonlatin_keypress](https://github.com/IgnorantGuru/spacefm/commit/6b698838d4ed46876c04568045e2e4a4c055e5c6#diff-c25c987f2f817e676da0f21cb5565eaaR390).  I probably won't get into this since they are atypical key shortcuts, but if you or someone else wants to investigate the event->group relationship, or provide a patch which handles those for ru, I'll consider including the adjustment.

The key branch has been merged into next.  My plans to convert non-latin keycodes in the config on upgrade doesn't look possible.  When you press a key, the transpose function uses the event->group and other event states to determine the latin equivalent.  These are not stored in the config - only the keycode is - and are thus unavailable for the conversion.  So no changes to the config are planned.

This means that some non-latin key shortcuts set in 1.0.2 and prior may not work correctly in 1.0.3 (next).  When you press the key, it will be converted, but the conversion will not match the key shortcut you had previously set.  So you may need to set new key shortcuts for those items.  This will be announced with the release.   Unfortunate upgrade breakage, but I don't see any other way to do it.

Thanks for testing.

**Comment by Vladimir-csp** [comment]
Thank you for fixing this!
Even with minor issues left, the greatest pain with shortcuts and layouts is gone. Hooray!

**Comment by IgnorantGuru** [comment]
@Vladimir-csp I think I found a way to convert the <= 1.0.2 key shortcuts as you use them.  This means when upgrading, any non-latin key shortcuts that were set should continue to work.  However, when downgrading they will break.  I think this is preferable.  It may not be perfect, but it should handle most situations okay.  Where it fails, you may need to set the shortcut.

I'll probably remove this code a few versions down the road, since it's not so efficient to search for both keycodes, but this should ease the upgrade process.  You might want to make sure there's no lag when typing nonlatin keys quickly, eg in the path bar or search.

If you still have your <= 1.0.2 config with some non-latin shortcuts, you can try upgrading and see if they work okay.  You'll also see it changing the keycodes if you diff the session file.

**Comment by IgnorantGuru** [comment]
Redo - in 236e3a7 I think I have a way for both upgrades and downgrades to stay intact, at least for key shortcuts set in 1.0.2 and prior.

Instead of converting the keycodes in the config, it just matches non-latin keycodes found in the config as well.  So older shortcuts should work.

Non-latin shortcuts added in 1.0.3 (next) and later probably won't work when downgrading.

**Comment by Vladimir-csp** [comment]
I never used non-latin shortcuts.
Ping @bitbybit

**Comment by bitbybit** [comment]
Wow. It's working! :+1:
Everything seems perfect. Will try to test more.

**Comment by bitbybit** [comment]
Just quick note that after month of using 1.0.2+next I did not find any troubles with shortcuts. As for non-latin combinations - they are translated to latin analogue in shortcut setup dialogue. For example if I set Ctrl + Г then it appears in menu as Ctrl + U. Both variants are working. As mentioned before combinations with special characters are working only in latin layout. I have no non-latin shortcuts from previous spacefm versions and actually I have no any customized shortcuts at all so I can not test if they are working in next version or not. So thanks again for fixing this.

**Comment by IgnorantGuru** [comment]
Thanks for testing.

**Comment by IgnorantGuru** [comment]
Made a change here in next today for GTK3, as the function use changed and was producing a warning about a NULL keymap.  Any testing to ensure this conversion is working with GTK3 appreciated.

**Comment by bitbybit** [comment]
Previously I have tried GTK2 version. Yes, GTK3 is also working.
